### PR TITLE
use the return value of assign to set the value of _in_experiment Fixes #73

### LINKED ISSUE #73: Make it easier to enable / disable exposure logging
The PlanOut language has a `return` construct that allows experimenters to programmatically enable / disable exposure logging.  For example, an experiment in which 20% of the users are eligible and are randomized into A and B can be written as:

```
eligible = bernoulliTrial(p=0.2, unit=userid);
if(!eligible) {
  return false;
}
variant = uniformChoice(choices=['a','b'], unit=userid);
```

However, many users of PlanOut do not use the PlanOut language and instead use a native API.  Modify the PlanOut reference implementation so that that the return value of the assign() method determines logging.  This would involve checking the return value of assign; if it is None (so that no return was triggered), in_experiment should be set to True; otherwise the truthiness of the return value should be used to keep track of the exposures.  These changes might also be integrated into PlanOut.js (cc @rawls238 ) and the Go implementation (cc @tsujeeth)

### PR DESCRIPTION
**Pull request by rawls238**
@eytan 

Where should we add this in the documentation?

### CODE REVIEW

File: python/planout/experiment.py [hunk 1]
```diff
@@ -69,7 +70,11 @@ def __init__(self, **inputs):
     def _assign(self):
         """Assignment and setup that only happens when we need to log data"""
         self.configure_logger()  # sets up loggers
-        self.assign(self._assignment, **self.inputs)
+
+        #consumers can optionally return False from assign if they don't want exposure to be logged
+        assign_val = self.assign(self._assignment, **self.inputs)
+        if self._in_experiment is self._default_in_experiment_val:
```
Comment (rawls238) [inline python/planout/experiment.py]: this still allows consumers to manually set self._in_experiment in the assign function as is done in the tests and only overrides it if _in_experiment is not set to the default value.

**Comment by eytan** [comment]
Thanks for implementing this @rawls238!  A few things:
1. I would like for users to be able to use anything with positive or negative truthiness, so perhaps we could do something like `self._in_experiment = True if assign_val or assign_val is None else False` should do the trick.
2. The _in_experiment attribute is undocumented private variable that is not part of the public API so I am not so concerned about backwards compatibility with respect to its usage, and would instead favor extreme clarity about the conductions under which randomized exposure data are logged (i.e. (1)).
3. Pretty sure exposure log control doesn't work correctly in SimpleInterpretedExperiment.  We probably need to insert `self._in_experiment = interpreterInstance.in_experiment` after experiment.py#L318.
4. Enabling / disabling exposure logging should probably be discussed in a subsection of planout/docs/logging.html on the github.io page. You can take a stab at this, but I would also be happy to do this.

Since this is a potentially breaking change and highlights an important aspect of the PlanOut language's semantics, I would also want to bump the version of PlanOut, update the `NEWS` file, and write a short highlighting the change.

**Comment by eytan** [comment]
Thanks for implementing this @rawls238!  A few things:
1. I would like for users to be able to use anything with positive or negative truthiness to control logging, so perhaps we could do something like `self._in_experiment = True if assign_val or assign_val is None else False` instead of testing for False return values directly?
2. The `_in_experiment` attribute is undocumented private variable that is not part of the public API so I am not so concerned about backwards compatibility with respect to its usage, and would instead favor extreme clarity about the conductions under which exposure data are logged (i.e. (1)).
3. Pretty sure exposure log control doesn't work correctly in SimpleInterpretedExperiment -- it seems that there should be a `self._in_experiment = interpreterInstance.in_experiment` after experiment.py#L318.
4. Re: documentation, Enabling / disabling exposure logging should probably be discussed in a subsection of planout/docs/logging.html on the github.io page. You are welcome to take a stab at this in a separate diff (it requires modifying a different branch), but I would be happy to write this up as well.
5. Since this is a potentially breaking change that relates to an important aspect of the PlanOut semantics, can you also bump the PlanOut version number in the Python package, and update the `NEWS` file?

**Comment by rawls238** [comment]
@eytan just to make sure we're on the same page, wondering what you think the main purpose of the return value should be? My original thought was that this should be used as a way to optionally short-circuit exposure logging by having the user force to return False, but it would be nice to have it be able to also turn logging on in certain cases so I'll make the change accordingly. 

Also, looking at the Simple Interpreted Experiment I agree that it does not look like it works properly with regard to the return operator. I'll fix it to work properly and I will update all the relevant documentation

**Comment by rawls238** [comment]
@eytan this should be ready to go, perhaps we should hold off releasing this until we implement https://github.com/facebook/planout/issues/74?

**Comment by eytan** [comment]
The main purpose of the return value is to control logging. In our
experience, eligibility checks go hand in hand with short circuiting
behavior, and if a parameter value is not determined by random assignment,
it shouldn't be logged -- and that is why it's natural to combine short
circuiting and logging via return.

I would be on board with implementing the other task before bumping the
planout version number.

**Comment by eytan** [comment]
Merged in, thanks @rawls238!

**Pull request merged**